### PR TITLE
Add hub filter for logging method invoke and error

### PR DESCRIPTION
**Pull request by frenzibyte**
Continues from #91 to further improve logging (mostly on the failure side).

Sample from `multiplayer.log` on a server failing to invoke `JoinRoomWithPassword(2, "test")` due to not being connected to the database:
```cs
2021-12-07 10:09:38 [verbose]: [user:1] Connected
2021-12-07 10:09:38 [verbose]: [user:2] Connected
2021-12-07 10:09:38 [verbose]: [user:3] Connected
2021-12-07 10:09:38 [verbose]: [user:4] Connected
2021-12-07 10:09:38 [verbose]: [user:5] Connected
2021-12-07 10:09:45 [verbose]: [user:1] Invoking hub method: JoinRoomWithPassword(2, "test")
2021-12-07 10:09:45 [verbose]: [user:1] Attempting to join room 2
2021-12-07 10:09:45 [error]: [user:1] Failed to invoke hub method: JoinRoomWithPassword(2, "test")
2021-12-07 10:09:45 [error]: MySqlConnector.MySqlException (0x80004005): Unable to connect to any of the specified MySQL hosts.
2021-12-07 10:09:45 [error]: at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, Int32 startTickCount, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 417
2021-12-07 10:09:45 [error]: at MySqlConnector.Core.ConnectionPool.ConnectSessionAsync(String logMessage, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 376
2021-12-07 10:09:45 [error]: at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 107
2021-12-07 10:09:45 [error]: at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 137
2021-12-07 10:09:45 [error]: at MySqlConnector.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 863
2021-12-07 10:09:45 [error]: at MySqlConnector.MySqlConnection.OpenAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 414
2021-12-07 10:09:45 [error]: at MySqlConnector.MySqlConnection.Open() in /_/src/MySqlConnector/MySqlConnection.cs:line 380
2021-12-07 10:09:45 [error]: at osu.Server.Spectator.Database.DatabaseAccess.getConnection() in /Users/salman/Desktop/osu-server-spectator/osu.Server.Spectator/Database/DatabaseAccess.cs:line 34
2021-12-07 10:09:45 [error]: at osu.Server.Spectator.Database.DatabaseAccess..ctor() in /Users/salman/Desktop/osu-server-spectator/osu.Server.Spectator/Database/DatabaseAccess.cs:line 22
2021-12-07 10:09:45 [error]: at osu.Server.Spectator.Database.DatabaseFactory.GetInstance() in /Users/salman/Desktop/osu-server-spectator/osu.Server.Spectator/Database/DatabaseFactory.cs:line 8
2021-12-07 10:09:45 [error]: at osu.Server.Spectator.Hubs.MultiplayerHub.JoinRoomWithPassword(Int64 roomId, String password) in /Users/salman/Desktop/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 41
2021-12-07 10:09:45 [error]: at lambda_method47(Closure , Object )
2021-12-07 10:09:45 [error]: at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.ExecuteMethod(ObjectMethodExecutor methodExecutor, Hub hub, Object[] arguments)
2021-12-07 10:09:45 [error]: at osu.Server.Spectator.LoggingHubFilter.InvokeMethodAsync(HubInvocationContext invocationContext, Func`2 next) in /Users/salman/Desktop/osu-server-spectator/osu.Server.Spectator/LoggingHubFilter.cs:line 27
```

I'm a bit unsure on my approach with how I'm logging from `LoggingHubFilter`, but didn't want to think too much about this.

I've also added method invoke logging, including the arguments provided using simply-written logic for that. Examples:
 - `Invoke(nameof(JoinRoomWithPassword), 2, "test")` logs `Invoking hub method: JoinRoomWithPassword(2, "test")`
 - `Invoke(nameof(TransferHost), 5)` logs `Invoking hub method: TransferHost(5)`
 - `Invoke(nameof(ChangeSettings), ...)` logs `Invoking hub method: ChangeSettings(Name:Unnamed room Password:no Type:TeamVersus Item:0 Queue:HostOnly)`
 - `Invoke(nameof(ChangeBeatmapAvailability), ...)` logs `Invoking hub method: ChangeBeatmapAvailability(Downloading, 75.00%)`
 - `Invoke(nameof(ChangeUserMods), ...)` logs `Invoking hub method: ChangeUserMods({ HD, DT (speed_change:1.05), CL })`

### CODE REVIEW

File: osu.Server.Spectator/Hubs/ILoggableHub.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Logging;
+
+namespace osu.Server.Spectator.Hubs
+{
+    internal interface ILoggableHub
```
Comment (bdach) [inline osu.Server.Spectator/Hubs/ILoggableHub.cs:9]: don't like the name, sounds wrong - this is a hub that can log (i.e. an `ILoggingHub`, or an `IHubWithLogging`, or even more abstractly an `ICanLog` or an `IHasLogger`), not a log that is *being* logged (which `ILoggableHub` implies)

File: osu.Server.Spectator/LoggingHubFilter.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+namespace osu.Server.Spectator
+{
+    /// <summary>
+    /// An <see cref="IHubFilter"/> logging method invoke and error to the <see cref="ILoggableHub"/>.
+    /// </summary>
+    public class LoggingHubFilter : IHubFilter
+    {
+        public async ValueTask<object?> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object?>> next)
+        {
+            var loggableHub = invocationContext.Hub as ILoggableHub;
```
Comment (bdach) [inline osu.Server.Spectator/LoggingHubFilter.cs:20]: would turn this into a guard to save on the repetitive checks, i.e.

```csharp
if (!(invocationContext.Hub is ILoggingHub loggingHub))
    return;
```

File: osu.Server.Spectator/Hubs/ILoggableHub.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Server.Spectator/Hubs/ILoggableHub.cs:9]: Was unsure about the name, funny enough I had `ILoggingHub` at first. Will rename to that.

File: osu.Server.Spectator/LoggingHubFilter.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Server.Spectator/LoggingHubFilter.cs:20]: I previously had `ILoggingHub` named as `IStatefulUserHub`, which made me concerned whether I want to do that. But now that I renamed it to `ILoggingHub`, would make sense to early-return now given that the entire hub filter is about logging.

**Comment by peppy** [comment]
Made a naming change and reorganised things to ensure this isn't too spammy on production.

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
Note that a lot of the parameters don't have a `ToString` implementation and show as empty strings right now. May be something we want to improve going forwards.

**Pull request merged**